### PR TITLE
Catch 2 test for Assert update viability

### LINKED ISSUE #2587: Create test for `analysis::assert_update_viable` in: update.cpp
Related to: #2574

### PR DESCRIPTION
**Pull request by sregales-TNO**
**Issue**
Resolves #2587 

**Approach**
Create a catch2 test for assert_update_viable

## Pre review checklist

- [x] make happy path test
- [x] make tests for the edge cases
- [x] leave code better than you found it

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

**Comment by sregales-TNO** [comment]
test this please

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+#include <catch2/catch.hpp>
+
+namespace analysis {
+bool assert_update_viable(const analysis_config_type *analysis_config,
+                          const state_map_type *source_state_map,
+                          const int ensemble_size,
+                          const local_updatestep_type *updatestep);
+}
+
+TEST_CASE("asserting_update_viable", "[analysis][prototype]") {
```
Comment (pinkwah) [inline libres/tests/analysis/test_assert_update_viable.cpp:19]: Nitpick: the function is called `assert_update_viable`
```suggestion
TEST_CASE("assert_update_viable", "[analysis][prototype]") {
```

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            REQUIRE(state_map_count_matching(source_state_map,
+                                             STATE_LOAD_FAILURE) == 1);
+
+            THEN("Assertion pass with realizations >= ensemble_size") {
+                REQUIRE(analysis::assert_update_viable(
+                    analysis_config, source_state_map, ensemble_size,
+                    updatestep));
+            }
+
+            ensemble_size += 1;
```
Comment (pinkwah) [inline libres/tests/analysis/test_assert_update_viable.cpp:90]: What is the value of `ensemble_size` at this point? Maybe prefer using a constant and comment? Ie:
```c++
func(..., 2 /* ensemble_size */)
```

**Review by pinkwah — COMMENTED** [review]
Tbh, not entirely sure what the purpose of `assert_update_viable` is, but the tests look fine enough.

File: libres/lib/enkf/analysis_config.cpp [hunk 3]
```diff
@@ -146,8 +146,8 @@ void analysis_config_set_max_runtime(analysis_config_type *config,
     config->max_runtime = max_runtime;
 }
 
-static void analysis_config_set_min_realisations(analysis_config_type *config,
-                                                 int min_realisations) {
+void analysis_config_set_min_realisations(analysis_config_type *config,
```
Comment (dafeda) [inline libres/lib/enkf/analysis_config.cpp:149]: Seems to me like this function is only used in `test_assert_update_viable.cpp` and in `analysis_config.cpp`.
If we are not to use it anywhere else, should we declare it in `test_assert_update_viable.cpp` instead of declaring it in a header?
This is what we've done with other "static" functions.

File: libres/lib/enkf/enkf_main.cpp [hunk 4]
```diff
@@ -556,8 +549,8 @@ static void enkf_main_add_ensemble_members(enkf_main_type *enkf_main) {
    is mainly to be able to test that the site config file is valid.
 */
 
-enkf_main_type *enkf_main_alloc(const res_config_type *res_config, bool strict,
-                                bool verbose, bool read_only) {
+enkf_main_type *enkf_main_alloc(const res_config_type *res_config, bool verbose,
```
Comment (dafeda) [inline libres/lib/enkf/enkf_main.cpp:552]: Could you write a sentence or two about what `strict` is and why it is no longer needed?

File: libres/lib/analysis/update.cpp [hunk 5]
```diff
@@ -550,10 +550,9 @@ Check whether the current state and config allows the update algorithm
 to be executed
 */
 bool assert_update_viable(const analysis_config_type *analysis_config,
-                          const enkf_fs_type *source_fs,
+                          const state_map_type *source_state_map,
```
Comment (dafeda) [inline libres/lib/analysis/update.cpp:553]: How come is it possible to replace `enkf_fs_type` with `state_map_type`?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    GIVEN("Allocated memory and default value parametrs") {
+        // alloc memory for parameters
+        analysis_config_type *analysis_config = analysis_config_alloc_default();
+        state_map_type *source_state_map = state_map_alloc();
+        local_updatestep_type *updatestep =
+            local_updatestep_alloc("name_not_important");
+        int ensemble_size = 1;
+        analysis_module_type *analysis_module =
+            analysis_module_alloc("not-important");
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:29]: I find it a bit surprising that its "not-important" what we call `analysis_module_alloc`.
Should this fail if called with an non-existant analysis-module @pinkwah ?

File: libres/lib/enkf/enkf_main.cpp (hunk 4 above)
Comment (sregales-TNO) [reply libres/lib/enkf/enkf_main.cpp:552]: strict is a parameter that was passed but never used/invoked. don't know if it's there to future proof the function or is a lingering parameter from it's past implementation. but eitherway it is not needed and might even confuse the use of the function. Since the documentation state a actual (non-existing) use of the parameter.

File: libres/lib/enkf/analysis_config.cpp (hunk 3 above)
Comment (sregales-TNO) [reply libres/lib/enkf/analysis_config.cpp:149]: I agree with this idea

**Comment by dafeda** [comment]
> Tbh, not entirely sure what the purpose of `assert_update_viable` is, but the tests look fine enough.

Are you talking about the purpose of the function `assert_update_viable` in `update.cpp`?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                    REQUIRE(!analysis::assert_update_viable( // change to True
+                        analysis_config, source_state_map, ensemble_size,
+                        updatestep));
+                }
+            }
+        }
+
+        // cleanup
+        local_ministep_free(ministep);
+        // analysis_module_free(analysis_module);
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:138]: Should the comments around this be removed?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        THEN("Assertion fail no active realization") {
+            REQUIRE(analysis_config_get_min_realisations(analysis_config) == 0);
+            REQUIRE(state_map_get_size(source_state_map) == 0);
+            REQUIRE(local_updatestep_get_num_ministep(updatestep) == 1);
+            REQUIRE(!analysis::assert_update_viable(
+                analysis_config, source_state_map, ensemble_size, updatestep));
+        }
+
+        WHEN("State is set to STATE_HAVE_DATA") {
+            state_map_iset(source_state_map, 0, STATE_INITIALIZED);
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:43]: Do we need to set `STATE_INITIALIZED` since we set `STATE_HAS_DATA` on the next line?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            THEN("Assertion pass") {
+                REQUIRE(analysis::assert_update_viable(
+                    analysis_config, source_state_map, ensemble_size,
+                    updatestep));
+            }
+        }
+
+        WHEN("Any state but STATE_HAS_DATA") {
+            realisation_state_enum states[3] = {
+                STATE_INITIALIZED, STATE_LOAD_FAILURE, STATE_PARENT_FAILURE};
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:57]: What about `STATE_UNDEFINED`?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            for (auto state : states) {
+                state_map_iset(source_state_map, 0, STATE_INITIALIZED);
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:59]: Do we need this given that state is set on next line?

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                state_map_iset(source_state_map, 0, state);
+                REQUIRE(state_map_get_size(source_state_map) == 1);
+                REQUIRE(state_map_count_matching(source_state_map, state) == 1);
+                THEN("Assertion fail no active realization") {
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:63]: I'm having a bit of trouble understanding this test.
It says that WHEN("Any state but STATE_HAS_DATA") but fails because there are no active realizations.
I can't see the connection between the state and the number of active realizations.

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                    REQUIRE(!analysis::assert_update_viable(
+                        analysis_config, source_state_map, ensemble_size,
+                        updatestep));
+                }
+            }
+        }
+
+        WHEN("one STATE_LOAD_FAILURE, but with others on STATE_HAS_DATA") {
+            for (int i = 0; i < 2; i++) {
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:72]: Could we do without this loop?

File: libres/lib/analysis/update.cpp (hunk 5 above)
Comment (sregales-TNO) [reply libres/lib/analysis/update.cpp:553]: Is not a replace perse, the change is done for two reasons. to reduce redundant function invokations and validation function should be in the business of observing not creating or fetching objects.
old implementation only uses `enkf_fs_type` to fetch a `state_map_type`, so if you already have an `enkf_fs_type` implies that you also already should have a `state_map_type` and `assert_update_viable doesn`t need to know what `enkf_fs_type` is.

furthermore when you look at the only use of `assert_update_viable` which is `smoother_update` you see that `enkf_fs_type` is beeing passed then `state_map_type` is being fetch for validation. And later after the validation, yet again `state_map_type` is being fetch for other uses. 

simplification: validation function will only observe, and move the retrieval of `state_map_type` before the validation.

File: libres/tests/analysis/test_assert_update_viable.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                state_map_iset(source_state_map, i, STATE_INITIALIZED);
```
Comment (dafeda) [inline libres/tests/analysis/test_assert_update_viable.cpp:73]: Ok, I'm starting to suspect that the state map first has to be set using `STATE_INITIALIZED` before doing anything else to it 😄 
Would you mind explaining how this works?

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 1 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:19]: I would go a bit further and say the fuction name should change, there is no assertion being done. it's more of a `validate_smoother_update`

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 2 above)
Comment (dafeda) [reply libres/tests/analysis/test_assert_update_viable.cpp:90]: I agree that it might be easier to read if it was a constant.

File: libres/lib/analysis/update.cpp (hunk 5 above)
Comment (dafeda) [reply libres/lib/analysis/update.cpp:553]: Sounds good 👍

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 1 above)
Comment (dafeda) [reply libres/tests/analysis/test_assert_update_viable.cpp:19]: What about `is_valid`? 
I think it makes OK sense since it's just used in `update.cpp`.

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 6 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:29]: this has been rectify locally should be either `"IES_ENKF" or "STD_ENKF"` but not yet pushed
And no it doesn't fail since it only looks at the `analysis_module_type->get_option` if `local_ministep_type > 1`

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 1 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:19]: winner winner chicken dinner

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 8 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:43]: Yes, and its because of the following https://github.com/equinor/ert/blob/a59aa5357653e83668f4cfabf474d3312a417b1c/libres/lib/enkf/state_map.cpp#L114

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 9 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:57]: will need to do that outside of the loop in a different `WHEN`, since `STATE_INITIALIZED` needs to be set prior to all of these states for it to be a legal transition. and this is not the case for `STATE_UNDEFINED`. or I could also add it to the list and put a if continue condition but that will kill the readablity of a test

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 10 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:59]: I'm only checking to make sure it's not an illegal state trasistion and was never change. But I see your point on the redundancy

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 11 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:63]: the validation function only counts realizations that has data, thus iff `STATE_HAS_DATA` any other state should result in a 0 and zero should always equate to a `false`, One thing I don't agree with is the stderr message

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 2 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:90]: ok agreed

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 7 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:138]: yeah we are still at the prototype, it's my lingering scaffolding

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 8 above)
Comment (dafeda) [reply libres/tests/analysis/test_assert_update_viable.cpp:43]: Thanks!

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 12 above)
Comment (dafeda) [reply libres/tests/analysis/test_assert_update_viable.cpp:72]: Yes, why a loop is a better question!

Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:72]: the for-loop is a lingering implementation, you can see from the code `WHEN` statement that I have 'others' meaning the list/loop was longer. but now that it's shorter there is no need for it. plus I think I should pull setting of state to a helper function to faciltate the construction and readablity of the tests

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 9 above)
Comment (sregales-TNO) [reply libres/tests/analysis/test_assert_update_viable.cpp:57]: `STATE_UNDEFINED` is the default and thats already checked at the paramaters initialization
https://github.com/equinor/ert/blob/a59aa5357653e83668f4cfabf474d3312a417b1c/libres/lib/enkf/state_map.cpp#L47

File: libres/lib/analysis/update.cpp [hunk 14]
```diff
@@ -566,10 +564,9 @@ bool assert_update_viable(const analysis_config_type *analysis_config,
         return false;
     }
 
-    // exit if multi step update
-    if ((local_updatestep_get_num_ministep(updatestep) > 1) &&
-        (analysis_config_get_module_option(analysis_config,
-                                           ANALYSIS_ITERABLE))) {
+    // exit if multi step update with iterable modules
```
Comment (sregales-TNO) [inline libres/lib/analysis/update.cpp:567]: Should a validation fuction have the behaviour/capability of exiting a program?

File: libres/tests/analysis/test_assert_update_viable.cpp (hunk 12 above)
Comment (dafeda) [reply libres/tests/analysis/test_assert_update_viable.cpp]: As for creating a helper function, I personally don't mind a bit of duplicated code in tests, but do go for it if you like.

**Comment by sregales-TNO** [comment]
The refactoring section has been moved to another PR #2853

**Comment by sregales-TNO** [comment]
Does any one know why the analysis_config_struct have both `analysis_module_type *analysis_module` and `std::unordered_map<std:string, analysis_module_type*> anlysis_modules`. I'm unable to determine the reason in documentation or code https://github.com/equinor/ert/blob/4c09ec6726883d9779cbdfe45eeaa45fc0481dcc/libres/lib/enkf/analysis_config.cpp#L48

**Comment by dafeda** [comment]
> Does any one know why the analysis_config_struct have both `analysis_module_type *analysis_module` and `std::unordered_map<std:string, analysis_module_type*> anlysis_modules`. I'm unable to determine the reason in documentation or code
> 
> https://github.com/equinor/ert/blob/4c09ec6726883d9779cbdfe45eeaa45fc0481dcc/libres/lib/enkf/analysis_config.cpp#L48

Have not checked the history so don't know who introduced this, but could you perhaps help us out here @joakim-hove ?

**Review by dafeda — APPROVED** [review]
LGTM 👍
